internal/testutils/testserver.go

```diff
@@ -51,55 +51,87 @@ func (t TestServer) FaultyClient() *http.Client {
 	return client
 }
 
-// HTTPMethod is an alias for string, representing an HTTP method.
-type HTTPMethod = string
-
-// ServerResponses is a map of HTTP methods to expected responses for testing.
-type ServerResponses map[HTTPMethod]struct {
-	ResponseCode        int                 // HTTP response status code.
-	ResponseBody        string              // HTTP response body.
-	ValidateRequestFunc func(*http.Request) // Function to validate incoming requests.
-}
-
 // NewHTTPTestServer creates a new HTTP test server with the specified responses for each HTTP method.
-func NewHTTPTestServer(t *testing.T, responses []ServerResponses) *TestServer {
+func NewHTTPTestServer(t *testing.T, responses []ResponseDef) *TestServer {
 	testServer := &TestServer{}
 	handler := func(rw http.ResponseWriter, req *http.Request) {
 		testServer.calls++
-		// Special case: if only one response is specified, this will be reused
-		// for each and every call
-		if len(responses) == 1 {
-			if res, found := responses[0][req.Method]; found {
-				if res.ValidateRequestFunc != nil {
-					res.ValidateRequestFunc(req)
-				}
-				rw.WriteHeader(res.ResponseCode)
-				_, _ = rw.Write([]byte(res.ResponseBody)) // nosemgrep: go.lang.security.audit.xss.no-direct-write-to-responsewriter.no-direct-write-to-responsewriter
-			} else {
-				t.Errorf("unexpected HTTP method call: %s", req.Method)
-			}
-			return
+		if len(responses) <= testServer.calls-1 {
+			t.Errorf("Exceeded number of calls to test server (expected: %d)", len(responses))
 		}
 
-		if len(responses) <= testServer.calls-1 {
-			t.Fail()
-			t.Fatalf("Exceeded number of calls to test server (expected: %d)", len(responses))
+		responseDef := responses[testServer.calls-1]
+		methodFuncs := map[string]func(*testing.T, *http.Request) Response{
+			http.MethodGet:    responseDef.GET,
+			http.MethodPost:   responseDef.POST,
+			http.MethodPut:    responseDef.PUT,
+			http.MethodDelete: responseDef.DELETE,
 		}
 
-		if res, found := responses[testServer.calls-1][req.Method]; found {
-			if res.ValidateRequestFunc != nil {
-				res.ValidateRequestFunc(req)
-			}
-			rw.WriteHeader(res.ResponseCode)
-			_, _ = rw.Write([]byte(res.ResponseBody)) // nosemgrep: go.lang.security.audit.xss.no-direct-write-to-responsewriter.no-direct-write-to-responsewriter
-		} else {
-			t.Errorf("unexpected HTTP method call: %s", req.Method)
+		handlerFunc, found := methodFuncs[req.Method]
+		if !found {
+			rw.WriteHeader(http.StatusMethodNotAllowed)
+			_, _ = rw.Write([]byte("Method not supported")) // nosemgrep: go.lang.security.audit.xss.no-direct-write-to-responsewriter.no-direct-write-to-responsewriter
+			return
 		}
+		if handlerFunc == nil {
```

Comment:
isn't that other than a strange edge-case the same as `!found` ? 
Why handle no response function being found as a HTTP error, and explicitly configuring nil for a method as a test error?

Reply:
:D yes, should we treat `!found` just as a test error?

Reply:
I'd say it makes sense, yes. 
If I didn't define that the server should reply to a PUT, and it gets an expected one, it would be reasonable to see that as a failing test IMO